### PR TITLE
22063 - Allow multiple payment methods, also create multiple CFS accounts to support EFT/PAD switching

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #:*
https://github.com/bcgov/entity/issues/22063

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/models/base_model.py [hunk 1]
```diff
@@ -60,4 +60,6 @@ def rollback():
     @classmethod
     def find_by_id(cls, identifier: int):
         """Return model by id."""
-        return db.session.get(cls, identifier)
+        if identifier:
+            return db.session.get(cls, identifier)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/base_model.py:64]: warning using .get on null depreciated

File: pay-api/src/pay_api/services/eft_short_names.py [hunk 2]
```diff
@@ -291,7 +291,7 @@ def get_pending_payment_count():
                 .filter(InvoiceModel.payment_account_id == PaymentAccountModel.id)
                 .filter(EFTCreditInvoiceLinkModel.status_code.in_([EFTCreditInvoiceStatus.PENDING.value]))
                 .correlate(PaymentAccountModel)
-                .as_scalar())
+                .scalar_subquery())
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_short_names.py:294]: warning as_scalar depreciated in sqlalchemy 1.4

File: pay-api/src/pay_api/services/statement.py [hunk 3]
```diff
@@ -172,6 +172,7 @@ def get_statement_eft_transactions(statement: StatementModel) -> List[EFTTransac
             .join(EFTCreditModel, EFTCreditModel.eft_transaction_id == EFTTransactionModel.id) \
             .join(EFTCreditInvoiceLinkModel, EFTCreditInvoiceLinkModel.eft_credit_id == EFTCreditModel.id) \
             .join(StatementInvoicesModel, StatementInvoicesModel.invoice_id == EFTCreditInvoiceLinkModel.invoice_id) \
+            .join(StatementModel, StatementModel.id == StatementInvoicesModel.statement_id) \
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:175]: Cartesian product without having this join

**Review by ochiu — APPROVED** [review]
Looks good

**Comment by seeker25** [comment]
thanks Ody, unit tests + lint + CI fixes to go

**Comment by seeker25** [comment]
This doesn't work in practice - didn't receive any feedback, thus we're going to try a different approach with multiple "sites"

**Comment by seeker25** [comment]
Have to change this entire PR so it creates a different site per CFS payment method

**Comment by seeker25** [comment]
Will reopen

**Comment by seeker25** [comment]
Note: Need to do a few tests for this:

1. Switching between BCOL / EFT / PAD
2. Switching between ONLINE_BANKING / CREDIT CARD

**Comment by seeker25** [comment]
Some unit test fixes, plus some local testing to go

File: pay-api/migrations/versions/2024_07_08_f6990cf5ddf1_cfs_account_payment_method.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'f6990cf5ddf1'
+down_revision = '0672573574f6'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    with op.batch_alter_table('cfs_accounts', schema=None) as batch_op:
```
Comment (seeker25) [inline pay-api/migrations/versions/2024_07_08_f6990cf5ddf1_cfs_account_payment_method.py:20]: Need migration for calling CFS site to get receipt_method

File: pay-api/src/pay_api/services/statement.py [hunk 5]
```diff
@@ -175,13 +175,14 @@ def get_payment_methods_query(statement: StatementModel):
             .select_from(subquery)
             .where(
                 and_(
+                    subquery.c.start_date != subquery.c.end_date,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:178]: Not sure if this is enough on the real system yet

File: pay-api/src/pay_api/services/statement.py [hunk 6]
```diff
@@ -163,8 +163,8 @@ def get_payment_methods_query(statement: StatementModel):
             select(union_query.c.id,
                    union_query.c.payment_method,
                    union_query.c.version,
-                   func.lag(union_query.c.end_date)
-                   .over(order_by=union_query.c.version.asc())
+                   func.coalesce(func.lag(union_query.c.end_date)
+                                 .over(order_by=union_query.c.version.asc()), datetime.min.date())
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:167]: Add in a default for start_date, being the min date

File: pay-api/src/pay_api/services/statement.py [hunk 7]
```diff
@@ -138,7 +138,7 @@ def get_payment_methods_query(statement: StatementModel):
                 PaymentAccountModel.id,
                 PaymentAccountModel.payment_method,
                 PaymentAccountModel.version,
-                literal(None).label('end_date')
+                literal(datetime.max.date()).label('end_date')
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:141]: default for end_date

File: pay-api/tests/unit/services/test_statement.py [hunk 8]
```diff
@@ -702,11 +702,22 @@ def test_statement_various_payment_methods_history(db, app):
     """Unit test to test various payment methods over the life of a statement."""
     # We aren't using the session fixture here because we want to test the history_cls
     # history_cls wont work on scoped session flush.
+    # Freezegun etc doesn't work here.
+    changed_column = [
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_statement.py:706]: Not super fun here...

**Review by Jxio — APPROVED** [review]
LGTM

**Comment by seeker25** [comment]
Fixing pay jobs, then local testing

File: pay-api/migrations/versions/2024_07_11_248cf4155e19_cfs_payment_method_populate.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        cfs_accounts = CfsAccountModel.query.filter(CfsAccountModel.status.in_([CfsAccountStatus.PENDING_PAD_ACTIVATION.value])).all()
+        for cfs_account in cfs_accounts:
+            cfs_account.payment_method = PaymentMethod.PAD.value
+            cfs_account.flush()
+
+        db.session.commit()
+
+def upgrade():
+    remove_versioning()
+    cfs_accounts = CfsAccountModel.query.options(
```
Comment (seeker25) [inline pay-api/migrations/versions/2024_07_11_248cf4155e19_cfs_payment_method_populate.py:108]: save the time decrypting bank account

**Pull request merged**